designs/0005-static-matrices.md

```diff
@@ -0,0 +1,95 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a `static_matrix` and `static_vector` type in the Stan language which on the backend will allow for significant performance opportunities. In Stan Math this will be represented by `var<Eigen::Matrix<double, 1, 1>>` or `var<matrix_cl<double>>`. The implementation will be fully backwards compatible with current Stan code. With optimizations from the compiler the conditions for a `matrix` to be a `static_matrix` can be detected and applied to Stan programs automatically for users.
```

Comment:
So this is also extending stan language? That was not previously clear form the issue. In that case I would propoese to add `const` keyword instead. That would be consistent with C++.

Reply:
I don't think they're quite `const`, because you could reassign a `static_matrix` (or `static_vector`) variable, as long as the thing being assignedis itself a static matrix (or vector).

Reply:
What about `static_row_vector`? 

Also, this is going to get confusing when we have `complex_matrix`, because then we'll have `matrix`, `static_matrix`, `complex_matrix`, and `static_complex_matrix`.  So I'm also wondering like @t4c1 if there's some way we can make this more combinatorial at the syntax level.  Maybe something like `static matrix `.  Then we could apply the same concept to arrays.

Reply:
> So this is also extending stan language? That was not previously clear form the issue. In that case I would propoese to add const keyword instead. That would be consistent with C++.

I'm half and half on extending the language for this. I wouldn't mind having this as something that strictly the compiler would look for and automate

> What about static_row_vector?

Yes I cover this below. Maybe something like 

```stan
fixed complex matrix[N, M] A = // var<Eigen::Matrix<complex<double>, -1, -1>>
```

Reply:
> I don't think they're quite const, because you could reassign a static_matrix (or static_vector) variable, as long as the thing being assigned is itself a static matrix (or vector).

How would you feel about making them const tho? We could change the definition I wrote above so that you can't assign after the first declaration.

Reply:
I think that would be just an arbitrary restriction. In other words no reason to do it.

Reply:
> In other words no reason to do it

Well there is a reason to do it -- it makes the naming simple :P.

But I think this breaks with arrays.

```
const matrix[N, M] matrices[P];
```

Is there an autodiff-themed name for this? We're really doing it for autodiff purposes.

Reply:
I like `fixed_matrix`, but that implies a new type rather than a new annotation.  I think that'll be simpler in the end, even if it may seem like more combinatorics to us (I say "seem" because I'm not sure it will be more combinatorics in the implementation given that "const" is very peculiar in meaning here).

Reply:
Hmm, maybe there needs to be an annotation/attribute design doc before we decide on this. We can use that for complex and sparse. That would probably be a big enough language feature to do the official stanc3